File: C++/代码随想录/Greedy/Sequential_problem/53. Maximum Subarray.cpp
// 53. Maximum Subarray
// https://leetcode.com/problems/maximum-subarray/

#include <iostream>
#include <vector>

int maxSubArray(const std::vector<int>& nums) {
    int currentMax = 0;
    int globalMax = 0;

    for (const auto& num : nums) {
        currentMax = std::max(num, currentMax + num);
        globalMax = std::max(globalMax, currentMax);
    }

    return globalMax;
}

int main() {
    std::vector<int> nums = {-2,1,-3,4,-1,2,1,-5,4};
    std::cout << maxSubArray(nums) << std::endl;

    std::vector<int> nums2 = {1};
    std::cout << maxSubArray(nums2) << std::endl;

    std::vector<int> nums3 = {5,4,-1,7,8};
    std::cout << maxSubArray(nums3) << std::endl;

    return 0;
}

File: C++/代码随想录/Greedy/Sequential_problem/53. Maximum Subarray.py
def maxSubArray(nums):
    if not nums:
        return 0

    max_sum = current_sum = nums[0]
    for num in nums[1:]:
        current_sum = max(num, current_sum + num)
        max_sum = max(max_sum, current_sum)

    return max_sum

# 测试用例
print(maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))  # 应该输出 6
print(maxSubArray([1]))                      # 应该输出 1
print(maxSubArray([5,4,-1,7,8]))             # 应该输出 23
